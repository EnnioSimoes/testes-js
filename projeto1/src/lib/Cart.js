import { remove } from 'lodash';
import find from 'lodash/find';

export default class Cart {
  items = [];

  add(item) {
    const itemToFind = { product: item.product };
    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind);
    }

    this.items.push(item);
  }

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price;
    }, 0);
  }

  remove(product) {
    remove(this.items, { product });
  }

  sumary() {
    const total = this.getTotal();
    const items = this.items;

    return {
      total,
      items,
    };
  }

  checkout() {
    const { total, items } = this.sumary();

    this.items = [];

    return {
      total,
      items,
    };
  }
}
