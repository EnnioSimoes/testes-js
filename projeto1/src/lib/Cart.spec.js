import Cart from './Cart.js';

describe('Cart', () => {
  let cart;
  let product = {
    title: 'TÃªnis Adidas',
    price: '35388',
  };

  beforeEach(() => {
    cart = new Cart();
  });

  it('should return 0 when getTotal() is executed in a newly instance', () => {
    expect(cart.getTotal()).toEqual(0);
  });

  it('should multiply quantity and price and receive the total amount', () => {
    const item = {
      product,
      quantity: 2,
    };

    cart.add(item);

    expect(cart.getTotal()).toEqual(70776);
  });

  it('should multiply quantity and price and receive the total amount', () => {
    cart.add({
      product,
      quantity: 2,
    });

    cart.add({
      product,
      quantity: 1,
    });

    expect(cart.getTotal()).toEqual(35388);
  });
});
