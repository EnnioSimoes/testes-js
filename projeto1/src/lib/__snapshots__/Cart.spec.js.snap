// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": "35388",
        "title": "Tênis Adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": "41872",
        "title": "Tênis Adidas AllStar",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;
